dist/ExtUtils-ParseXS/lib/perlxs.pod

```diff
@@ -1,2065 +1,4706 @@
+builds, the macros gracefully degrade to a single global instance. These
+macros have C<MY_CXT> ("my context") as part of their names.
 
-Although primarily designed for use with multi-threaded Perl, the macros
-have been designed so that they will work with non-threaded Perl as well.
+It is therefore strongly recommended that these macros be used by all XS
+modules that make use of static data.
 
-It is therefore strongly recommended that these macros be used by all
-XS modules that make use of static data.
+When creating a new skeleton F<Foo.xs> file, you can use the C<--global>
+option of F<h2xs> to also include a skeleton set of macros, e.g.
 
-The easiest way to get a template set of macros to use is by specifying
-the C<-g> (C<--global>) option with h2xs (see L<h2xs>).
+    h2xs -A --global -n Foo::Bar
 
-Below is an example module that makes use of the macros.
+Below is a complete example module that makes use of the macros. It tracks
+the names of up to three blind mice.
 
     #define PERL_NO_GET_CONTEXT
     #include "EXTERN.h"
     #include "perl.h"
     #include "XSUB.h"
+    #include "ppport.h"
+
+    #define MAX_NAME_LEN 100
 
     /* Global Data */
 
     #define MY_CXT_KEY "BlindMice::_guts" XS_VERSION
 
     typedef struct {
         int count;
-        char name[3][100];
+        char name[3][MAX_NAME_LEN+1];
     } my_cxt_t;
 
     START_MY_CXT
 
-    MODULE = BlindMice           PACKAGE = BlindMice
+    MODULE = BlindMice  PACKAGE = BlindMice
+
+    PROTOTYPES: DISABLE
 
     BOOT:
     {
         MY_CXT_INIT;
         MY_CXT.count = 0;
-        strcpy(MY_CXT.name[0], "None");
-        strcpy(MY_CXT.name[1], "None");
-        strcpy(MY_CXT.name[2], "None");
     }
 
     int
-    newMouse(char * name)
-        PREINIT:
-          dMY_CXT;
-        CODE:
-          if (MY_CXT.count >= 3) {
-              warn("Already have 3 blind mice");
-              RETVAL = 0;
-          }
-          else {
-              RETVAL = ++ MY_CXT.count;
-              strcpy(MY_CXT.name[MY_CXT.count - 1], name);
-          }
-        OUTPUT:
-          RETVAL
+    AddMouse(char *name)
+      PREINIT:
+        dMY_CXT;
+      CODE:
+        if (strlen(name) > MAX_NAME_LEN)
+            croak("Mouse name too long\n");
+
+        if (MY_CXT.count >= 3) {
+            warn("Already have 3 blind mice");
+            RETVAL = 0;
+        }
+        else {
+            RETVAL = ++MY_CXT.count;
+            strcpy(MY_CXT.name[MY_CXT.count - 1], name);
+        }
+      OUTPUT:
+        RETVAL
 
     char *
-    get_mouse_name(index)
-          int index
-        PREINIT:
-          dMY_CXT;
-        CODE:
-          if (index > MY_CXT.count)
-            croak("There are only 3 blind mice.");
-          else
+    get_mouse_name(int index)
+      PREINIT:
+        dMY_CXT;
+      CODE:
+        if (index > MY_CXT.count)
+            croak("There are only %d blind mice.", MY_CXT.count);
+        else
             RETVAL = MY_CXT.name[index - 1];
```

Comment:
"index" is a weird name for this arg since you're really saying "give me the first, second, or third mouse.."

And if you pass 0 we end up reading ```[index - 1] !```